From 8dc2f71193965bfc60474696a420532e90a56626 Mon Sep 17 00:00:00 2001
From: Yang Wei <Wei.Yang@windriver.com>
Date: Tue, 14 Jan 2014 19:05:30 +0800
Subject: [PATCH 1/2] powerpc/mm: Fix processor is stuck on e6500 rev1

In tlb_unlock depends R7 and R8, currently these two register
are corrupted. so we use R0 and R9.

Signed-off-by: Yang Wei <Wei.Yang@windriver.com>
---
 arch/powerpc/mm/tlb_nohash_low.S |   18 ++++++++++--------
 1 files changed, 10 insertions(+), 8 deletions(-)

diff --git a/arch/powerpc/mm/tlb_nohash_low.S b/arch/powerpc/mm/tlb_nohash_low.S
index 961c443..0f03570 100644
--- a/arch/powerpc/mm/tlb_nohash_low.S
+++ b/arch/powerpc/mm/tlb_nohash_low.S
@@ -329,10 +329,11 @@ ALT_MMU_FTR_SECTION_END_IFCLR(MMU_FTR_USE_TLBILX)
 
 #ifdef CONFIG_FSL_ERRATUM_A_004801
 .macro tlb_lock
-    mfspr   r7, SPRN_PVR
-	lis r8, PVR_VER_E6500
-	ori r8, r8, PVR_REV_E6500V1
-	cmpw r7, r8
+	mfspr   r0, SPRN_PVR
+	rlwinm r0, r0, 0, 24, 15
+	lis r9, PVR_VER_E6500
+	ori r9, r9, PVR_REV_E6500V1
+	cmpw r0, r9
 	bne 1f
 	ld	r7,PACA_TLB_PER_CORE_PTR(r13)
 	mtocrf	0x01,r7
@@ -354,10 +355,11 @@ ALT_MMU_FTR_SECTION_END_IFCLR(MMU_FTR_USE_TLBILX)
 .endm
 
 .macro tlb_unlock
-    mfspr   r7, SPRN_PVR
-	lis r8, PVR_VER_E6500
-	ori r8, r8, PVR_REV_E6500V1
-	cmpw r7, r8
+	mfspr   r0, SPRN_PVR
+	rlwinm r0, r0, 0, 24, 15
+	lis r9, PVR_VER_E6500
+	ori r9, r9, PVR_REV_E6500V1
+	cmpw r0, r9
 	bne 1f
 	mtocrf	0x01,r7
 	bf	31,1f		/* no lock if TLB_PER_CORE_HAS_LOCK clear */
-- 
1.7.5.4

